File: LeetCode/3. 无重复字符的最长子串.cpp
//
// Created by sgh on 2019-03-23.
//

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0;
        int res = 0;
        map<char, int> record;
        for (int i = 0; i < s.size(); i ++) {
            if (record.find(s[i]) != record.end() && record[s[i]] >= left) {
                res = max(res, i - left);
                left = record[s[i]] + 1;
            } else if (i == s.size() - 1) {
                res = max(res, i - left + 1);
            }
            record[s[i]] = i;
        }
        return res;
    }
};

File: LeetCode/3. 无重复字符的最长子串.go
package LeetCode

func lengthOfLongestSubstring(s string) int {
	var (
		res    = 0
		l, r   = 0, 0
		record = map[byte]int{}
	)

	max := func(a, b int) int {
		if a > b {
			return a
		}

		return b
	}

	for ; r < len(s); r++ {
		if i, ok := record[s[r]]; ok && i >= l {
			res = max(res, r-l)
			l = i + 1
		} else if r == len(s)-1 {
			res = max(res, r-l+1)
		}

		record[s[r]] = r
	}

	return res
}
